corehq/ex-submodules/casexml/apps/phone/restore.py

```diff
@@ -706,31 +666,30 @@ def _get_task_timeout(self, new_task):
 
     def _generate_restore_response(self, async_task=None):
         """
-        This function returns a RestoreResponse class that encapsulates the response.
+        :returns: A file-like object containing response content.
         """
         username = self.restore_user.username
         count_items = self.params.include_item_count
-        with self.restore_state.restore_class(username, count_items) as response:
+        with RestoreContent(username, count_items) as content:
             for provider in get_element_providers(self.timing_context):
                 with self.timing_context(provider.__class__.__name__):
-                    response.extend(provider.get_elements(self.restore_state))
+                    content.extend(provider.get_elements(self.restore_state))
 
             for provider in get_async_providers(self.timing_context, async_task):
                 with self.timing_context(provider.__class__.__name__):
-                    provider.extend_response(self.restore_state, response)
+                    provider.extend_response(self.restore_state, content)
 
-        return response
+            return content.get_fileobj()
 
-    def set_cached_payload_if_necessary(self, resp, duration):
-        cache_payload_path = resp.name
+    def set_cached_payload_if_necessary(self, fileobj, duration, async):
         # on initial sync, only cache if the duration was longer than the threshold
         is_long_restore = duration > timedelta(seconds=INITIAL_SYNC_CACHE_THRESHOLD)
 
-        if self.force_cache or is_long_restore or self.sync_log:
-            self._set_cache_in_redis(cache_payload_path)
-
-    def _set_cache_in_redis(self, cache_payload_path):
-        self.restore_payload_path_cache.set_value(cache_payload_path, self.cache_timeout)
+        if async or self.force_cache or is_long_restore or self.sync_log:
```

Comment:
I think `self.force_cache` is currently always set to `True` if `self.async = True`. 

https://github.com/dimagi/commcare-hq/blob/f9d71d0c13857ea5342ca7343a359859d9a775e6/corehq/ex-submodules/casexml/apps/phone/restore.py#L513

Was that not getting set correctly when the task is running?

I think if you want to add the async check explictly here, it would be good to remove it from the force_cache, otherwise it might get a little confusing.

Reply:
> Was that not getting set correctly when the task is running?

No, I just didn't realize the two were related.

> I think if you want to add the async check explictly here, it would be good to remove it from the force_cache, otherwise it might get a little confusing.

Good point. dc91b42c436b483a930eae005d1d4bc0f5aefb7a